002-es6-modules.md

```diff
@@ -0,0 +1,274 @@
+    - A hook for when an import is exactly performed.
+    
+* [CreateImportBinding](https://tc39.github.io/ecma262/#sec-createimportbinding)
+    - A means to create a shared binding (variable) from an export to an import. Required for the "live" binding of imports.
+
+* [ModuleNamespaceCreate](https://tc39.github.io/ecma262/#sec-modulenamespacecreate)
+    - Provides a means of creating a list of exports manually, used so that CommonJS `module.exports` can create ModuleRecords that are prepopulated.
+
+---
+
+### [WhatWG Loader](https://github.com/whatwg/loader)
+
+---
+
+Discusses the design of module metadata in a [Registry](https://whatwg.github.io/loader/#registry). All actions regarding the Registry are synchronous.
+
+**NOTE** It is not Node's intent to implement the asynchronous pipeline in the Loader specification. There is discussion about including a synchronous pipeline in the specification as an addendum.
+
+
+### [Summary Video on Youtube](youtube.com/watch?v=NdOKO-6Ty7k)
+
+## Additional Structures Required
+
+### DynamicModuleRecord
+
+A Module Record that presents a view of an Object for its `[[Namespace]]` rather than coming from an environment record.
+
+The list of exports is frozen upon construction. No new properties may be added. No properties may be removed.
+
+## Algorithm
+
+When `require()`ing a file.
+
+1. Determine if file is ES6 or CommonJS (CJS).
+2. If CJS
+	1. Evaluate immediately
+	2. Produce a DynamicModuleRecord from `module.exports`
+3. If ES6
+	1. Parse for `import`/`export`s and keep record, in order to create bindings
+	2. Gather all submodules by performing `require` recursively
+		* See circular dep semantics below 
+	3. Connect `import` bindings for all relevant submodules (see [ModuleDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-moduledeclarationinstantiation))
+	4. Evaluate
+
+## Semantics
+
+### Determining if source is an ES6 Module
+
+If a module can be parsed outside of the ES6 Module goal, it will be treated as a [Script](https://tc39.github.io/ecma262/#sec-parse-script). Otherwise it will be parsed as [ES6](https://tc39.github.io/ecma262/#sec-parsemodule).
+
+In pseudocode, it looks somewhat like:
+
+```
+try {
+  v8::Script::Compile
+}
+catch (e) {
+  v8::Script::CompileModule
+}
+```
+
+V8 may or may not choose to use parser fallback to combine this into one step.
+
+### CommonJS consuming ES6
+
+#### default exports
+
+ES6 modules only ever declare named exports. A default export just exports a property named `default`.
+
+Given
+
+```javascript
+let foo = 'my-default';
+export default foo;
+```
+
+```javascript
+require('./es6');
+// {default:'my-default'}
+```
+
+#### read only
+
+The objects create by an ES6 module are [ModuleNamespace Objects](https://tc39.github.io/ecma262/#sec-module-namespace-exotic-objects).
+
+These have `[[Set]]` be a no-op and are read only views of the exports of an ES6 module.
+
+### ES6 consuming CommonJS
```

Comment:
> Do you know which npm modules you use are CoffeeScript, TypeScript, PureScript,

I think languages that compile to JS (CoffeeScript, TypeScript) and module systems (CJS, ES modules) are two different and not that relevant matters.
Also even in that case, people when using some library, naturally by learning/using it, just know on which modules system it stands or in which language (if not direct JS) it's written.

> Thats a HUGE change that would break pretty much all modules

How come? We all rely on semver, nobody blindly switches major version of dependency.

>> so as long some module is written as CJS compilation step is mandatory.
> Can you clarify?

I referred to browser environments. If you would want to natively use an ES module that relies on CJS module (even if required via `import`) then this CJS module with al its CJS dependencies needs to be recompiled to ES modules to have initial ES module working.

> but having ambiguous code is worse.

Wouldn't it be more ambiguous to have CJS modules required via `import`, or ES modules imported via `require` ?

Letting `require` into ES modules, makes thing way simpler, no need for new file extension, no need for translation of two formats, and probably some other things, which now you struggle to solve.